doc/decisions/0_drafts/transformations.md

```diff
@@ -0,0 +1,128 @@
+
+```
+/DISPLAY/BRIGHTNESS = 100
+``` 
+
+As can be seen, the keys are in UPPERCASE within the configuration file.
+In Elektra keys are case-sensitive. 
+As operations on keysets such as `ksLookup` operate with *runtime* data after the post-storage phase, `kdb get /DISPLAY/BRIGHTNESS` will fail.
+For Elektra, the key `/DISPLAY/BRIGHTNESS` does not exist, as the `rename` plugin transformed this into the lowercase `/display/brightness`.
+
+This leads to problems with the notification plugins.
+As notification plugins are executed after the post-storage phase of the `set` operation, they will receive a keyset with the already transformed keys.
+In this example, the notification plugins will receive all-UPPERCASE keys, and send out notifications with those all-UPPERCASE keys. 
+An application listening to those notifications will not be able to query Elektra for those keys, as for Elektra those UPPERCASE keys do not exist.
+
+Apart from the problems with notifications, the way key name changing plugins work also breaks change tracking in plugins like `dbus`.
+This is because key names are read-only when they are contained in a `KeySet`.
+In order to change the name of a key, such a plugin has to create a new key with the changed name, and delete the key with the old name.
+The `dbus` plugin implements change tracking by checking the 'key needs sync' flag instead of comparing the values.
+As new keys by design have the 'key needs sync' flag set, the plugins that implement change tracking via the flag will always erroneously detect transformed keys as changed.
+
+### Observed problems with changing values (normalization)
+
+Suppose we have a plugin that changes the value of a key to the hard-coded value `1` during the storage phase of the `set` operation.
+If a plugin does change tracking, this will lead to false positives.
+
+```
+user:/limits/openfiles = 1
+```
+
+If the user changes the value, e.g. using `kdb set user:/limits/openfiles 23`, plugins will observe the new value `23`.
+However, the value-changing plugin will reset that value back to `1`.
+So in practice, the configuration has not been changed.
+Plugins relying on change tracking plugins (e.g. notification plugins) will however think that it has.
+
+### Observed problems with adding and removing metadata
+
+This is a bit of a mixture between changing key names and changing values.
+
+## Constraints
+
+1.
+2.
+3.
+
+## Assumptions
+
+1. We want to enable some kind of transformations
+2.
+3.
+
+## Considered Alternatives
+
+### Enable setting and unsetting of the `keyNeedsSync` flag
+
+Require plugins that rename keys to remove the `keyNeedsSync` flag.
+This would require making the 'key sync' APIs public.
+This does not solve the problem of changing values.
+
+### Use change tracking to replace the `keyNeedsSync` flag
+
+Eliminate the need of the `keyNeedsSync` flag by utilizing the planned change tracking API.
+This does not solve the problem of changing values and false positives.
```

Comment:
I would actually make it an assumption that `keyNeedsSync` is replaced with the new change tracking implementation.

Reply:
On second thought, replacing this flag entirely is not a good idea. Detecting changes will always be more compute-intensive than just checking a flag. This might be overkill for plugins that do not really care about the changes.

Reply:
I am not so sure if there is a big difference. strcmp on such little values is very fast. Comparing 4/8 byte values is **maybe** nearly as fast as checking a flag (benchmark needed).

But I agree with @atmaxinger: if we already have an optimization, there should be a good reason to throw it.

Reply:
Okay maybe making it an assumption is too much. But replacing `keyNeedsSync` should still be an option we explore. However, I think it's mostly independent of what we discuss here.

I brought it up, because the `keyNeedsSync` flag is actually not as straightforward as it seems. It doesn't cover all cases and even in `libelektra-kdb` (which has more information than e.g. plugins) we need to do extra work. IMO `keyNeedsSync` may still be useful for `libelektra-kdb`, but shouldn't be public API and plugins/applications should use the new change tracking instead.